### PR TITLE
Fix OS X issue with some string methods.

### LINKED ISSUE #7132: BUG: String tokenizing methods return garbled content with some UTF-8 characters
## Problem description

We stumbled across this, when trying to `json_encode` some content from SilverStripe and the method suddenly returned nothing (although `json_encode` should be capable of encoding _any_ proper UTF-8 string). 

The issue only arises when using methods that work on portions of strings, such as `HTMLText::FirstSentence`, `HTMLText::Summary` etc. When running content through some of these methods, you might end up with invalid UTF-8.

### Affected Versions

SilverStripe Framework 2.x - 4.x
String Fieldtype classes (`Text`, `HTMLText`,…), `Convert` class, maybe the shortcode-parser.

### Severity

Edge-case, but has a high WTF-factor when it's happening.

## Source of the issue

The methods that work on substrings use methods that are not multibyte safe. In my tests, I found out that `preg_replace` and `preg_split` on whitespace (`/\s+/`) can split a multibyte character in half, which will result in invalid UTF-8.

The problem only seems to occur if your system uses an UTF-8 version of a locale.

## Illustration of the problem

```php
// Must set an UTF-8 variant of a locale
setlocale(LC_ALL, "en_US.UTF-8");
// Generate a non-breaking space, U+00A0
$str = html_entity_decode('&nbsp;',ENT_COMPAT, 'UTF-8' );
// replace whitespace
echo preg_replace('/\s+/', ' ', $str);
```

This will output non-valid UTF-8, you'll see the dreaded: �

### Explanation

When using an UTF-8 locale, `preg_replace` seems to detect the byte `A0` as white-space and replace it with the byte `20`. So the byte sequence `C2A0` (which is the non-breaking space), becomes `C220`, which is _not_ a valid UTF-8 character. It's very likely, that other UTF-8 characters could be affected by this problem as well (not only the non-breaking space).

## Possible fix

All instances of `preg_replace('/\s+/', …)` and `preg_split('/\s+/',…)` should be replaced with the multibyte-safe variants, `mb_ereg_replace` and `mb_split`. Other string operations should be looked into, preferably all string replacements and regular expressions should operate in a multibyte-safe manner.

### PR DESCRIPTION
**Pull request by bummzack**
Fix OS X issue with `Convert::html2raw`, `HTMLText::FirstSentence`, `HTMLText::Summary` and `Text::FirstSentence`.
Use unicode modifier for regular expressions that deal with whitespace.
Added unit-tests to ensure no invalud utf-8 gets generated by these methods.
Fixes #7132

### CODE REVIEW

File: tests/core/ConvertTest.php [hunk 1]
```diff
@@ -396,4 +396,30 @@ public function testBase64url() {
...
+			$this->markTestIncomplete('Unable to run this test because of missing locale!');
+			return;
+		}
+
+		$problematicText = html_entity_decode('<p>This is a&nbsp;Test with non-breaking&nbsp;space!</p>', ENT_COMPAT, 'UTF-8');
+
+		$this->assertTrue(mb_check_encoding(Convert::html2raw($problematicText), 'UTF-8'));
+
+		// reset locale
+		setlocale(LC_ALL, $currentLocale);
```
Comment (tractorcow) [inline tests/core/ConvertTest.php]: This reset needs to be in tearDown(), not in the test itself. A failed test will not reset the locale correctly.

File: tests/model/HTMLTextTest.php [hunk 2]
```diff
@@ -314,4 +314,34 @@ public function testShortCodeParsedInTemplateHelpers() {
...
+		$problematicText = html_entity_decode('<p>This is a&nbsp;Test with non-breaking&nbsp;space!</p>', ENT_COMPAT, 'UTF-8');
+
+		$textObj = new HTMLText('Test');
+		$textObj->setValue($problematicText);
+
+		$this->assertTrue(mb_check_encoding($textObj->FirstSentence(), 'UTF-8'));
+		$this->assertTrue(mb_check_encoding($textObj->Summary(), 'UTF-8'));
+
+		// reset locale
+		setlocale(LC_ALL, $currentLocale);
```
Comment (tractorcow) [inline tests/model/HTMLTextTest.php]: As above, please use tearDown() to reset.

File: tests/core/ConvertTest.php (hunk 1 above)
Comment (tractorcow) [inline tests/core/ConvertTest.php]: Also please see http://php.net/manual/en/function.setlocale.php#106811; setlocale(LC_ALL, 0) can return multiple locales.

**Comment by tractorcow** [comment]
Note : This PR will need to be merged up to 4 branch manually.

File: tests/core/ConvertTest.php (hunk 1 above)
Comment (bummzack) [inline tests/core/ConvertTest.php]: @tractorcow I see. Should I implement the locale set/reset as outlined in that user-comment? Would it hurt to set/clear the locale for all tests in these classes? Eg. run `setlocale` in `setUp` and reset in `tearDown`?

Comment (bummzack) [inline tests/core/ConvertTest.php]: I've updated the tests with a tearDown implementation. Changed `LC_ALL` to `LC_CTYPE`, as this setting is enough for the test in question and prevents getting multiple locales from `setlocale`.
